GH-2818: DSL support for -ws module

garyrussell opened the pull request:
Resolves https://github.com/spring-projects/spring-integration/issues/2818

Docs to follow tomorrow.

garyrussell committed:
GH-2818: DSL support for -ws module

Resolves https://github.com/spring-projects/spring-integration/issues/2818

artembilan commented on spring-integration-ws/src/main/java/org/springframework/integration/ws/AbstractWebServiceOutboundGateway.java:
```diff
@@ -165,6 +165,12 @@ public void setFaultMessageResolver(FaultMessageResolver faultMessageResolver) {
+	 */
+	@Deprecated
 	public void setMessageSender(WebServiceMessageSender messageSender) {
 		Assert.state(!this.webServiceTemplateExplicitlySet,
```
Would you mind to change logic then in favor of calling a replacing API instead?
Then it is going to be much easier to support possible changes in the target logic afterward.

garyrussell replied:
Sorry, I missed this; but, that said, I don't understand it.

artembilan replied:
I mean this:
```
public void setMessageSender(WebServiceMessageSender messageSender) {
   setMessageSenders(messageSender);
}
```

So, we don't need to have that `Assert` in this method any more. The logic from a deprecated method is fully delegated to replacing API.

garyrussell replied:
Doh; ok.

artembilan commented on spring-integration-ws/src/main/java/org/springframework/integration/ws/dsl/BaseWsOutboundGatewaySpec.java:
```diff
@@ -0,0 +1,151 @@
+	 * @return the spec.
+	 * @see org.springframework.web.util.UriComponentsBuilder
+	 */
+	public S encodeUri(boolean encodeUri) {
```
Hm. Looks like we need to consider to support an `EncodingMode` instead. Similar to what we have now in HTTP and WebFlux.

Different issue though

garyrussell replied:
True (different issue).

artembilan replied:
https://github.com/spring-projects/spring-integration/issues/3180

artembilan commented on spring-integration-ws/src/main/java/org/springframework/integration/ws/dsl/MarshallingWsInboundGatewaySpec.java:
```diff
@@ -0,0 +1,75 @@
+	 * @param marshaller the marshaller.
+	 * @param unmarshaller the unmarshaller.
+	 */
+	protected MarshallingWsInboundGatewaySpec(Marshaller marshaller, Unmarshaller unmarshaller) {
```
I wouldn't expose two of these ctors since it feels much natural just to call those configuration options...

artembilan commented on spring-integration-ws/src/main/java/org/springframework/integration/ws/dsl/MarshallingWsOutboundGatewaySpec.java:
```diff
@@ -0,0 +1,128 @@
+	 * @param unmarshaller the unmarshaller.
+	 * @param messageFactory the message factory.
+	 */
+	protected MarshallingWsOutboundGatewaySpec(String uri,
```
I think we need to consider to have all of these ctors spreaded as options and build the final `MarshallingWebServiceOutboundGateway` in the `doGet()`.
If possible, of course...

Otherwise it doesn't look too useful to have such a DSL spec based only on ctors

garyrussell replied:
Yeah - will rework - seemed a little clunky to me; I think a CTOR that takes a WST would be useful and return a different spec in that case (one that doesn't offer the properties that are disallowed with an explicit WST; there are many).

artembilan commented on spring-integration-ws/src/main/java/org/springframework/integration/ws/dsl/SimpleWsOutboundGatewaySpec.java:
```diff
@@ -0,0 +1,109 @@
+	 * @param sourceExtractor the source extractor.
+	 * @param messageFactory the message factory.
+	 */
+	protected SimpleWsOutboundGatewaySpec(String uri,
```
Same here.

We may live with what you have so far, but I wonder if we really can make it more natural builder style way...

artembilan reviewed (changes requested):
Some review to think about.

Thanks

garyrussell committed:
* Rework to improve fluency

- reduce the number of factory methods and use a different spec when
  an external WST is provided.

artembilan commented:
@garyrussell ,

any updates? Or do you want me to take over this from you?

Thanks

garyrussell commented:
? 

I pushed rework a couple of days ago - didn't you get a notification?

artembilan commented:
>  garyrussell force-pushed the garyrussell:GH-2818 branch from 2a96e81 to d4bdbc0 

Seems for me when we "force-pushed", there is no notification. GH flaw? 😉 

OK. Sorry for noise: I should check PR before bothering you.
Will review soon.

Thanks

garyrussell commented:
I know force-push for a rebase doesn't notify, but I assumed force push with additional commits would - but I guess not.

In future, I'll try to remember to push the new commit; then force-push after rebasing.

artembilan commented on spring-integration-ws/src/test/java/org/springframework/integration/ws/dsl/WsDslTests.java:
```diff
@@ -0,0 +1,215 @@
+ * @since 5.3
+ *
+ */
+public class WsDslTests {
```
Is it expected that we don't have any integration test with this new DSL used from an `IntegrationFlow`?

garyrussell replied:
I am not sure why an integration test is necessary since these tests verify the adapters are populated properly. But I can add one if you insist.

artembilan reviewed (changes requested):
Some non-critical review.

I can address them all on merge (including an integration test)

Thanks

P.S. Although there has to be some Docs on the matter! 😄

garyrussell commented on spring-integration-ws/src/main/java/org/springframework/integration/ws/dsl/BaseWsInboundGatewaySpec.java:
```diff
@@ -0,0 +1,65 @@
+
+	protected abstract E create();
+
+	protected E assemble(E gateway) {
```
Just for consistency with the outbound spec, in case common components are added later.

artembilan reviewed (changes requested):
OK. One more comment and code style. Plus docs and we are good to go.
Everything else is in agreement silently.

garyrussell committed:
* Docs and address comments

garyrussell commented:
Force pushed with additional commit after rebase for doc conflicts.
